File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-2.java
class Solution {
    public int findMin(int[] nums) {
        if (nums.length <= 1) {
            return nums[0];
        }
        int l = 0;
        int r = nums.length - 1;
        int min = nums[0];
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] < min) {
                min = nums[mid];
            }
            if (nums[mid] >= nums[l] && nums[mid] > nums[r]) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
        return min;
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        res = nums[0]
        prefix, suffix = 0, 0
        for i in range(len(nums)):
            prefix = nums[i] * (1 if prefix == 0 else prefix)
            suffix = nums[n - 1 - i] * (1 if suffix == 0 else suffix)
            res = max(res, prefix, suffix)
        return res
